pkg/manifest/manifest_loader.go

```diff
@@ -0,0 +1,296 @@
+		return Manifest{}, []error{fmt.Errorf("file `%s` is not yaml", manifestPath)}
+	}
+
+	data, err := afero.ReadFile(fs, manifestPath)
+
+	if err != nil {
+		return Manifest{}, []error{err}
+	}
+
+	return parseManifest(fs, manifestPath, data)
+}
+
+func parseManifest(fs afero.Fs, manifestPath string, data []byte) (Manifest, []error) {
+	manifestPath = filepath.Clean(manifestPath)
+	var manifest manifest
+
+	err := yaml.Unmarshal(data, &manifest)
+
+	if err != nil {
+		return Manifest{}, []error{err}
+	}
+
+	var errors []error
+
+	workingDir := filepath.Dir(manifestPath)
+
+	projectDefinitions, projectErrors := toProjectDefinitions(fs, workingDir, manifest.Projects)
+
+	if projectErrors != nil {
+		errors = append(errors, projectErrors...)
+	} else if len(projectDefinitions) == 0 {
+		errors = append(errors, fmt.Errorf("no projects defined in `%s`", manifestPath))
+	}
+
+	environmentDefinitions, manifestErrors := toEnvironments(fs, workingDir, manifest.Environments)
+
+	if manifestErrors != nil {
+		errors = append(errors, manifestErrors...)
+	} else if len(environmentDefinitions) == 0 {
+		errors = append(errors, fmt.Errorf("no environments defined in `%s`", manifestPath))
+	}
+
+	if errors != nil {
+		return Manifest{}, errors
+	}
+
+	return Manifest{
+		Projects:     projectDefinitions,
+		Environments: environmentDefinitions,
+	}, nil
+}
+
+func toEnvironments(fs afero.Fs, workingDir string, groups []group) (map[string]EnvironmentDefinition, []error) {
+	var errors []error
+	environments := make(map[string]EnvironmentDefinition)
+
+	for _, group := range groups {
+		var groupName string
+
+		if group.Group == nil {
+			groupName = DEFAULT_GROUP
+		} else {
+			groupName = *group.Group
+
+			if groupName == DEFAULT_GROUP {
+				return nil, []error{
+					fmt.Errorf("cannot add group with name `%s`. name is reserved", DEFAULT_GROUP),
+				}
+			}
+		}
+
+		for _, conf := range group.Entries {
+			env, configErrors := toEnvironment(conf, groupName)
+
+			if configErrors != nil {
+				errors = append(errors, configErrors...)
+				continue
+			}
+
+			if _, found := environments[env.Name]; found {
+				errors = append(errors, fmt.Errorf("environment with name `%s` already exists", env.Name))
+				continue
+			}
+
+			environments[env.Name] = env
+		}
+	}
+
+	if errors != nil {
+		return nil, errors
+	}
+
+	return environments, nil
+}
+
+func toEnvironment(config environment, group string) (EnvironmentDefinition, []error) {
+	var errors []error
+
+	token, err := parseToken(config, group, config.Token)
+
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	if config.Url == "" {
+		errors = append(errors, fmt.Errorf("url is missing or blank"))
+	}
+
+	if errors != nil {
+		return EnvironmentDefinition{}, nil
+	}
+
+	return EnvironmentDefinition{
+		Name:  config.Name,
+		Url:   config.Url,
+		Token: token,
+		Group: group,
+	}, nil
+}
+
+func parseToken(config environment, group string, token tokenConfig) (Token, error) {
```

Comment:
`config environment, group string, ` are unused params.

Reply:
Good point! 
There are still proper error messages missing in that file. So `group` and
`environment` will be needed to specify the location of the